src/seraphis_impl/enote_store.cpp

```diff
@@ -345,18 +408,39 @@ bool SpEnoteStore::try_import_legacy_key_image(const crypto::key_image &legacy_k
             "sp enote store (import legacy key image): intermediate and full legacy maps inconsistent (bug).");
 
         // c. set the full record
-        get_legacy_enote_record(record_ref->second.record,
-            legacy_key_image,
-            m_legacy_contextual_enote_records[legacy_enote_identifier].record);
-        events_inout.emplace_back(NewLegacyRecord{legacy_enote_identifier});
-
         // d. set the full record's contexts
-        update_contextual_enote_record_contexts_v1(
-                record_ref->second.origin_context,
-                spent_context,
-                m_legacy_contextual_enote_records[legacy_enote_identifier].origin_context,
-                m_legacy_contextual_enote_records[legacy_enote_identifier].spent_context
-            );
+        if (record_ref->second.is_type<LegacyContextualIntermediateEnoteRecordV1>())
+        {
+            m_legacy_contextual_enote_records[legacy_enote_identifier] = LegacyContextualEnoteRecordV1{};
+            get_legacy_enote_record(enote_record_ref(record_ref->second),
+                legacy_key_image,
+                m_legacy_contextual_enote_records[legacy_enote_identifier].unwrap<LegacyContextualEnoteRecordV1>().record);
+            events_inout.emplace_back(NewLegacyRecord{legacy_enote_identifier});
+
+            update_contextual_enote_record_contexts_v1(
+                    record_ref->second.unwrap<LegacyContextualIntermediateEnoteRecordV1>().origin_context,
+                    spent_context,
+                    m_legacy_contextual_enote_records[legacy_enote_identifier].unwrap<LegacyContextualEnoteRecordV1>().origin_context,
+                    m_legacy_contextual_enote_records[legacy_enote_identifier].unwrap<LegacyContextualEnoteRecordV1>().spent_context
+                );
+        }
+        else if (record_ref->second.is_type<LegacyContextualIntermediateEnoteRecordV2>())
+        {
+            m_legacy_contextual_enote_records[legacy_enote_identifier] = LegacyContextualEnoteRecordV2{};
+            get_legacy_enote_record(enote_record_ref(record_ref->second),
+                legacy_key_image,
+                m_legacy_contextual_enote_records[legacy_enote_identifier].unwrap<LegacyContextualEnoteRecordV2>().record);
+            events_inout.emplace_back(NewLegacyRecord{legacy_enote_identifier});
+
+            update_contextual_enote_record_contexts_v1(
+                    record_ref->second.unwrap<LegacyContextualIntermediateEnoteRecordV2>().origin_context,
+                    spent_context,
+                    m_legacy_contextual_enote_records[legacy_enote_identifier].unwrap<LegacyContextualEnoteRecordV2>().origin_context,
+                    m_legacy_contextual_enote_records[legacy_enote_identifier].unwrap<LegacyContextualEnoteRecordV2>().spent_context
+                );
+        }
+        else
+            CHECK_AND_ASSERT_THROW_MES(false, "LegacyContextualIntermediateEnoteRecord not implemented");
```

Comment:
I don't like this, suggestions how to make it more compact / less ugly are welcome.